plugins/anomaly-detection/AnomalyDetection.py

```diff
@@ -0,0 +1,594 @@
+def calculate_ewma_anomaly(dataframe):
+    """
+    The function categorizes whether a network flow is Anomalous or not
+    based on the calculated EWMA value.
+    A network flow record is anomalous if abs(throughput - ewma) > Standard
+    Deviation.
+    True - Anomalous Traffic, False - Not Anomalous
+
+    Args:
+        dataframe : The row of a dataframe containing all data related to
+        this network connection.
+    Returns:
+        A list of boolean values which signifies if a network flow record
+        of that connection is anomalous or not.
+    """
+    stddev = dataframe[7]
+    ewma_arr = dataframe[9]
+    throughput_arr = dataframe[10]
+    anomaly_result = []
+
+    if ewma_arr is None:
+        logger.error("Error: EWMA values not calculated for this flow record")
+        result = False
+        anomaly_result.append(result)
+    elif throughput_arr is None:
+        logger.error("Error: Throughput values not in ideal format for this "
+                     "flow record")
+        result = False
+        anomaly_result.append(result)
+    else:
+        num_records = len(ewma_arr)
+
+        for i in range(num_records):
+            if stddev is None:
+                logger.error("Error: Too Few Throughput Values for Standard "
+                             "Deviation to be calculated.")
+                result = False
+            elif throughput_arr[i] is None:
+                logger.error("Error: Throughput values not in ideal format "
+                             "for this flow record")
+                result = False
+            else:
+                float(stddev)
+                result = True if (abs(float(throughput_arr[i]) - float(
+                    ewma_arr[i])) > 
+                    float(stddev)) else False
+
+            anomaly_result.append(result)
+
+    return anomaly_result
+
+
+def calculate_arima(diff_secs_throughput):
+    """
+    The function calculates AutoRegressive Integrated Moving Average
+    (ARIMA) for a given list of throughput values of a connection
+    Assumption: Since ARIMA needs a handful of data to train and start
+    prediction, any connection with less than 3 flow records will not be
+    taken into account for calculation. We return empty value in that case.
+    Args:
+        diff_secs_throughput: Column of a dataframe containing difference
+        in seconds between connection start and current flow,
+        along with its throughput as a tuple
+    Returns:
+        A list of ARIMA values calculated for the set of throughput values
+        for that specific connection.
+    """
+
+    throughput_list = []
+    for ele in diff_secs_throughput:
+        throughput_list.append(float(ele[1]))
+
+    if len(throughput_list) <= 3:
+        logger.error("Error: Too Few throughput values for ARIMA to work with")
+        return None
+    else:
+        try:
+            warnings.filterwarnings("ignore")
+            throughput_list = [x + 1 for x in throughput_list]
```

Comment:
May I ask why we need to `+1` for each throughput and `-1` in `predictions_final`, is it required by ARIMA? Will doing this affect accuracy？

Reply:
there is a change which is minimal, without the +1 and -1 the anomaly is still detected correctly. removing it